orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +853,368 @@ def _update_onnx_model_initializers(self, state_tensors):
         for w_i in replace_indices:
             del self._onnx_model.graph.initializer[w_i]
         self._onnx_model.graph.initializer.extend(new_weights)
+
+    def _extract_model_states(self, state_dict, pytorch_format):
+        """Extract model states from the training session and load into the state_dict"""
+
+        model_states = self._training_session.get_model_state(include_mixed_precision_weights=False)
+        state_dict[_utils.state_dict_model_key()] = {}
+
+        # extract trained model weights from the training session
+        for precision in model_states:
+            state_dict[_utils.state_dict_model_key()][precision] = {}
+            for model_state_key in model_states[precision]:
+                if pytorch_format:
+                    state_dict[_utils.state_dict_model_key()][precision][model_state_key] = \
+                        torch.from_numpy(model_states[precision][model_state_key])
+                else:
+                    state_dict[_utils.state_dict_model_key()][precision][model_state_key] = \
+                        model_states[precision][model_state_key]
+
+        # extract untrained (frozen) model weights
+        precision_map = _utils.state_dict_precision_map()
+        for node in self._onnx_model.graph.initializer:
+            if node.name not in state_dict[_utils.state_dict_model_key()][precision] and \
+                node.name in self.options.utils.frozen_weights:
+                if node.data_type in precision_map:
+                    if pytorch_format:
+                        state_dict[_utils.state_dict_model_key()][precision_map[node.data_type]][node.name] = \
+                            torch.from_numpy(onnx.numpy_helper.to_array(node))
+                    else:
+                        state_dict[_utils.state_dict_model_key()][precision_map[node.data_type]][node.name] = \
+                            onnx.numpy_helper.to_array(node)
+
+    def _extract_optimizer_states(self, state_dict):
+        """Extract optimizer states from the training session and load into the state_dict"""
+
+        optimizer_states = self._training_session.get_optimizer_state()
+        state_dict[_utils.state_dict_optimizer_key()] = {}
+        for model_state_key in optimizer_states:
```

Comment:
we can maybe just do a deepcopy like `state_dict[_utils.state_dict_optimizer_key()] = copy.deepcopy(optimizer_states)`

Reply:
oh yeah. Initially I was converting to tensor, now no need to iterate over it. Will change this.